lightning/src/offers/invoice.rs

```diff
@@ -491,19 +494,48 @@ where
 
 impl UnsignedBolt12Invoice {
 	fn new(invreq_bytes: &[u8], contents: InvoiceContents) -> Self {
+		// TLV record ranges applicable to invreq_bytes.
+		const NON_EXPERIMENTAL_TYPES: core::ops::Range<u64> = 0..INVOICE_REQUEST_TYPES.end;
+		const EXPERIMENTAL_TYPES: core::ops::Range<u64> =
+			EXPERIMENTAL_OFFER_TYPES.start..EXPERIMENTAL_INVOICE_REQUEST_TYPES.end;
+
+		let (_, _, _, invoice_tlv_stream, _, _, experimental_invoice_tlv_stream) =
+			contents.as_tlv_stream();
+
+		let mut bytes = Vec::with_capacity(
+			invreq_bytes.len()
+				+ invoice_tlv_stream.serialized_length()
+				+ experimental_invoice_tlv_stream.serialized_length(),
+		);
+
 		// Use the invoice_request bytes instead of the invoice_request TLV stream as the latter may
 		// have contained unknown TLV records, which are not stored in `InvoiceRequestContents` or
 		// `RefundContents`.
-		let (_, _, _, invoice_tlv_stream) = contents.as_tlv_stream();
-		let invoice_request_bytes = WithoutSignatures(invreq_bytes);
-		let unsigned_tlv_stream = (invoice_request_bytes, invoice_tlv_stream);
+		for record in TlvStream::new(invreq_bytes).range(NON_EXPERIMENTAL_TYPES) {
+			record.write(&mut bytes).unwrap();
+		}
 
-		let mut bytes = Vec::new();
-		unsigned_tlv_stream.write(&mut bytes).unwrap();
+		invoice_tlv_stream.write(&mut bytes).unwrap();
 
-		let tagged_hash = TaggedHash::from_valid_tlv_stream_bytes(SIGNATURE_TAG, &bytes);
+		let mut experimental_bytes = Vec::with_capacity(
+			invreq_bytes.len()
+				- TlvStream::new(invreq_bytes)
+					.range(SIGNATURE_TYPES)
+					.last()
+					.map_or(0, |last_record| last_record.end)
+				+ experimental_invoice_tlv_stream.serialized_length(),
+		);
```

Comment:
Similarly here, this would seem clearer to me. Unless there are performance concerns that I'm missing, which seems possible. 
```rust
		let mut experimental_bytes = Vec::with_capacity(
			TlvStream::new(invreq_bytes)
				.range(EXPERIMENTAL_TYPES)
				.map(|record| record.serialized_length())
				.sum::<usize>()
			.saturating_add(experimental_invoice_tlv_stream.serialized_length())
		);

Reply:
Mostly performance concerns of serializing all the experimental TLVs. Would this be clearer?

```rust
		let mut experimental_bytes = Vec::with_capacity(
			invreq_bytes.len()
				- TlvStream::new(invreq_bytes)
					.range(EXPERIMENTAL_TYPES)
					.first()
					.map_or(invreq_bytes.len(), |first_record| first_record.start)
				+ experimental_invoice_tlv_stream.serialized_length(),
		);
```

Reply:
I think that is a bit clearer, though I don't think we expect that many experimental TLVs to be present so not sure about the performance concern there. Won't hold up the PR over this though.